File: desktop/ui/chat_commands.mjs
/**
 * Decisions behind the desktop composer's in-chat commands — the desktop half
 * of what `comrade_core::command` parses.
 *
 * ## Why this is its own module
 *
 * `main.js` is a classic script full of DOM glue and cannot be unit tested, so
 * the parts that *decide* something live here and `node --test` covers them —
 * the split `focus_view.mjs`, `together_sync.mjs` and `call_decisions.mjs`
 * already use. "It's only a couple of lines in the input handler" is exactly how
 * the rest of this UI became untestable.
 *
 * ## What is *not* here
 *
 * **The grammar.** Which words are commands, what `@handle` means, and how a
 * free-text query becomes a recording are all `comrade_core::command`'s answers,
 * reached over the Tauri bridge (`parse_chat_command`, `resolve_mentions`,
 * `play_query`). Nothing in this file re-parses composer text — a second grammar
 * is precisely the failure `/pay` already demonstrates, having drifted across
 * four implementations.
 *
 * What *is* here is the part that is genuinely the desktop's own: given a parsed
 * command, what should this window do, and what should it say — including the
 * cases where the honest answer is "not here yet".
 */

/** Nothing to do; send the text as typed. */
export const SEND = "send";
/** Route to `tara_aside`, never to `send_dm`. `/tara`. */
export const ASIDE = "aside";
/**
 * Route to `tara_in_chat` — `@tara`, which the peer sees.
 *
 * A separate action from {@link ASIDE} because the audience is different, and a
 * composer that treated them the same would either publish a private thought or
 * hide an answer the other person was waiting for.
 */
export const TARA_HERE = "tara_here";
/** Route to `assign_task`. */
export const TASK = "task";
/** Route to `offer_action`. */
export const OFFER = "offer";
/** Route to a `/play` flow. */
export const PLAY = "play";
/** Open a local screen. */
export const OPEN = "open";
/** Show the command list. */
export const HELP = "help";
/** Refuse, and say why — the command cannot work from this window. */
export const BLOCKED = "blocked";
/** Refuse, and say what is missing — the command is incomplete. */
export const INCOMPLETE = "incomplete";
/**
 * Two contacts answer to one `@handle`; ask which before doing anything.
 *
 * Not {@link INCOMPLETE}: that said "pick which one" and offered nothing to pick,
 * which is a dead end dressed as an instruction. This carries the candidates so
 * the composer can show them and re-run the same draft (see {@link withChoices}).
 */
export const CHOOSE = "choose";

/**
 * Screens this desktop build actually has.
 *
 * `breathe`, `focus`, `journal`, `read` and `tara` all exist on Android; on
 * desktop only Focus and the reader shipped (`docs/ATTENTION.md` §7,
 * `docs/FRONTEND_STRATEGY.md`'s divergence table). Listing what is *here*
 * rather than assuming parity is what lets `/journal` say "not on desktop yet"
 * instead of opening nothing and looking broken.
 */
export const DESKTOP_SCREENS = new Set(["focus", "read"]);

/**
 * Whether this window can *play* a together session.
 *
 * **True since 2026-08-05**, and the exit condition it was written against is
 * the one that was met: the `<video>` element, the file picker, and the
 * `play_query`/`play_route` commands are all in place, so `/play` can open a
 * session here rather than answering "there is no player here yet" — which had
 * become false while still being said. `desktop/ui/play_flow.mjs` decides what
 * each route does with it.
 *
 * Kept as a named constant rather than deleted: `PlayRoute::PlayEmbed` still
 * has nowhere to go in this window (no iframe — `docs/TOGETHER.md` §7 carries
 * the CSP condition), so "this window can play" remains a real question with a
 * real answer, and `planPlay` refuses that one route by name.
 */
export const DESKTOP_CAN_PLAY = true;

/**
 * What the composer should do with a parsed command.
 *
 * Takes the `ChatCommand` as the bridge delivers it (serde-tagged, so
 * `{ kind: "task", text, assignees }`) and returns
 * `{ action, message, ...payload }`. `message` is always present for the two
 * refusing actions and otherwise omitted.
 *
 * The one rule worth stating: **an unknown command never sends.** A message
 * silently delivered as `/frobnicate the thing` is a message the other person
 * gets confused by; a message silently *swallowed* is worse. So it is reported.
 */
export function planFor(command, { mentions = [] } = {}) {
  if (!command || typeof command.kind !== "string") return { action: SEND };
  switch (command.kind) {
    case "plain":
      return { action: SEND };

    case "help":
      return { action: HELP };

    case "pay":
      // `/pay` is recognised so it is not called a typo, but the composer sends
      // it as text exactly as it always has — `extract_payments` draws the chip
      // and the peer's own client detects the intent on receipt.
      return { action: SEND };

    case "ask_tara": {
      const text = (command.text || "").trim();
      if (!text) {
        return {
          action: INCOMPLETE,
          message: "Say what you want to think through — only you will see it.",
        };
      }
      return { action: ASIDE, text };
    }

    case "tara_here": {
      const text = (command.text || "").trim();
      if (!text) {
        return {
          action: INCOMPLETE,
          message: "Ask her something — you'll both see her answer.",
        };
      }
      return { action: TARA_HERE, text };
    }

    case "task": {
      const text = (command.text || "").trim();
      if (!text) {
        return { action: INCOMPLETE, message: "What needs doing?" };
      }
      const targets = resolvedTargets(command.assignees, mentions);
      if (targets.problem) return targets.problem;
      // No target is a note to self, which is the common case and needs no
      // resolution at all.
      return { action: TASK, text, peer: targets.npubs[0] ?? null };
    }

    case "offer_to": {
      const action = command.action;
      const targets = resolvedTargets(command.targets, mentions);
      if (!command.targets || command.targets.length === 0) {
        return {
          action: INCOMPLETE,
          message: "Name the comrade you want to offer this to, like @ana.",
        };
      }
      if (targets.problem) return targets.problem;
      return { action: OFFER, appAction: action, peers: targets.npubs };
    }

    case "open": {
      const key = command.action;
      if (!DESKTOP_SCREENS.has(key)) {
        return {
          action: BLOCKED,
          message: `${labelFor(key)} is on the phone app, not here yet.`,
        };
      }
      return { action: OPEN, appAction: key };
    }

    case "play":
      if (!DESKTOP_CAN_PLAY) {
        return {
          action: BLOCKED,
          message: "Listening together works on the phone app; there is no player here yet.",
        };
      }
      return { action: PLAY, query: command.query || "", service: command.service ?? null };

    case "unknown":
      return {
        action: INCOMPLETE,
        message: `There is no /${command.name} — type / to see what there is.`,
      };

    default:
      // A command this build does not know about: send nothing, say nothing
      // wrong. Reached when the Rust grammar gains a variant before this file
      // does, which is a release-skew case rather than a bug.
      return { action: SEND };
  }
}

/**
 * Turn parsed mentions into npubs, or explain why we cannot.
 *
 * The ambiguous case is the one that matters: two contacts can answer to one
 * handle, because a handle is a self-declared alias and anyone may publish any
 * name. Picking one is how a private message reaches the wrong person, so it
 * comes back as a question.
 */
function resolvedTargets(parsed, mentions) {
  const npubs = [];
  for (const m of parsed || []) {
    const match = mentions.find((r) => r.handle === m.handle);
    if (!match || (!match.npub && (!match.candidates || match.candidates.length === 0))) {
      return {
        npubs,
        problem: {
          action: INCOMPLETE,
          message: `@${m.handle} isn't in your contacts.`,
        },
      };
    }
    if (!match.npub) {
      // The answer exists; only the user has it. One handle at a time — a draft
      // naming two ambiguous handles would otherwise raise a chooser that can be
      // half-answered, and the next one is asked about on the re-run.
      return {
        npubs,
        problem: {
          action: CHOOSE,
          handle: m.handle,
          candidates: match.candidates,
          message: `Two contacts answer to @${m.handle} — which one?`,
        },
      };
    }
    npubs.push(match.npub);
  }
  return { npubs, problem: null };
}

/**
 * Every action key the bridge can send, i.e. `AppAction::key()` in Rust.
 *
 * Exported so a test can assert {@link labelFor} covers all of them. These used
 * to be spelled `breath` here and `breathe` on the wire, and because `labelFor`
 * falls back to the raw key the only symptom was a sentence reading "breathe is
 * on the phone app" — the kind of drift a `/breath/i` assertion happily passes.
 */
export const ACTION_KEYS = ["breathe", "focus", "journal", "tara", "read"];

/** Human name for an action key, for the "not here yet" sentences. */
export function labelFor(key) {
  return (
    {
      breathe: "Taking a deep breath",
      focus: "Focus sessions",
      journal: "The journal",
      tara: "Tara",
      read: "The reader",
    }[key] || key
  );
}

/**
 * The `/` picker's rows for what has been typed so far.
 *
 * Matches on the canonical name *and* every alias, so typing `/listen` finds
 * `play` — otherwise an alias that works when submitted would be invisible while
 * typing, which reads as it not existing.
 *
 * Returns `null` when the picker should be closed rather than empty: an empty
 * dropdown hanging under the composer while somebody types an ordinary message
 * containing a slash is noise.
 */
export function pickerRows(text, catalog) {
  if (typeof text !== "string" || !text.startsWith("/")) return null;
  // Once there is a space, the command has been chosen and the picker's job is
  // done — what follows is an argument.
  if (/\s/.test(text)) return null;
  const typed = text.slice(1).toLowerCase();
  const rows = (catalog || []).filter(
    (spec) =>
      spec.name.startsWith(typed) ||
      (spec.aliases || []).some((a) => a.startsWith(typed)),
  );
  return rows.length ? rows : null;
}

/**
 * The text to put in the composer when a picker row is chosen.
 *
 * A trailing space when the command takes an argument, so the caret is already
 * where the next word goes; none when it takes nothing, so pressing Enter
 * submits rather than sending `/breathe `.
 */
export function completionFor(spec) {
  if (!spec || !spec.name) return "";
  const needsArgument = Boolean(spec.argument) || Boolean(spec.takes_mention);
  return needsArgument ? `/${spec.name} ` : `/${spec.name}`;
}

/** `/tara` — nothing leaves this device. */
export const TARA_PRIVATE = "private";
/** `@tara` — the question and her answer both go to the peer. */
export const TARA_SHARED = "shared";

/**
 * Which Tara audience the raw draft implies, or `null` if she is not addressed.
 *
 * Called on every keystroke, on the raw text rather than a parse, because it must
 * be right from the moment `@tara ` is typed — before there is anything to parse.
 * The sigil is the whole difference between a private thought and a message the
 * other person reads, so the composer has to label it before Enter is pressed.
 */
export function taraDraft(text) {
  if (typeof text !== "string") return null;
  const t = text.trimStart();
  const lower = t.toLowerCase();
  for (const [prefix, audience] of [
    ["@tara", TARA_SHARED],
    ["/tara", TARA_PRIVATE],
  ]) {
    if (lower === prefix) return audience;
    if (lower.startsWith(prefix) && /\s/.test(t.charAt(prefix.length))) return audience;
  }
  return null;
}

/**
 * What a shared Tara answer carries on the wire — the mirror of
 * `comrade_core::tara::TARA_CHAT_PREFIX`.
 *
 * Only needed because a live-arriving DM reaches this frontend as the raw wire
 * body: `messages_with` already hands back `{author, content}` with the marker
 * off, but `incoming_direct_message` does not go through that DTO. Both paths
 * feed the same `state.dms`, so without this mirror a message would read one way
 * as it arrived and another way after a reload.
 */
export const TARA_CHAT_PREFIX = "Tara: ";

/**
 * Split a raw wire body into `{author, content}`, exactly as
 * `comrade_ui::split_author` does.
 *
 * **Attribution, not attestation.** Anybody can type "Tara: ", so a match means
 * the sending Comrade says this came from her — see `AUDIT.md` Q17. Style a
 * bubble with it; never trust it.
 */
export function splitAuthor(content) {
  const text = typeof content === "string" ? content : "";
  return text.startsWith(TARA_CHAT_PREFIX)
    ? { author: "tara", content: text.slice(TARA_CHAT_PREFIX.length) }
    : { author: "human", content: text };
}

/**
 * Apply the choices a user has made for ambiguous handles, so re-running the
 * same draft reaches the person they picked.
 *
 * A choice is honoured **only** when it names one of that handle's own
 * candidates. Without that check a pin left over from an earlier draft — or from
 * before a contact was removed — would silently retarget a message, which is the
 * exact failure the ambiguity exists to prevent.
 */
export function withChoices(mentions, chosen) {
  const picks = chosen || {};
  return (mentions || []).map((match) => {
    const pick = picks[match.handle];
    const offered = (match.candidates || []).some((c) => c.npub === pick);
    return !match.npub && pick && offered ? { ...match, npub: pick } : match;
  });
}

/**
 * How one candidate reads in the chooser.
 *
 * The key is not decoration: two contacts answering to one handle very often
 * share an alias too — that is *why* the handle is ambiguous — so a row showing
 * only the name would offer two identical choices. `title` is the caller's
 * resolved display name; the shortened key is what tells them apart.
 */
export function candidateLabel(title, npub) {
  const key = typeof npub === "string" && npub.length > 16
    ? `${npub.slice(0, 10)}…${npub.slice(-4)}`
    : npub || "";
  return `${title} · ${key}`;
}

/**
 * The line under the composer for a rendered offer or task bubble, so a received
 * one gets an affordance rather than reading as a bare sentence.
 *
 * Parsing is the bridge's (`comrade_core::command::parse_offer_line` /
 * `karya::parse_task_line`); this only decides whether *this* window can act on
 * the result. An offer of a screen desktop does not have is shown without a
 * button rather than hidden — the sentence is still the point, and hiding a
 * comrade's message because our build lacks a screen would be the wrong trade.
 */
export function offerAffordance(appAction) {
  if (!appAction) return null;
  return {
    label: labelFor(appAction),
    actionable: DESKTOP_SCREENS.has(appAction),
  };
}

File: android/app/src/main/java/mullu/comrade/ui/ChatCommands.kt
package mullu.comrade.ui

import mullu.comrade.ComradeCore
import uniffi.comrade_core.AppAction
import uniffi.comrade_core.ChatCommand
import uniffi.comrade_core.CommandSpec
import uniffi.comrade_core.MusicService
import uniffi.comrade_ui.PlayRoute

/**
 * Decisions behind the chat composer's in-chat commands — the Android half of
 * what `comrade_core::command` parses.
 *
 * Kept free of any Compose or Android import so the whole thing is unit-testable
 * on a plain JVM, exactly like [ComposerMode], [ChatThread] and
 * [mullu.comrade.voice.VoiceCommand]. The desktop twin is
 * `desktop/ui/chat_commands.mjs` and the two are pinned by mirrored test
 * vectors — where they disagree on purpose (which screens exist), each says so.
 *
 * ## What is *not* here
 *
 * **The grammar.** Which words are commands, what `@handle` means, and how free
 * text becomes a recording are all Rust's answers, reached through
 * [ComradeCore.parseChatCommand] / [ComradeCore.resolveMentions]. Nothing in
 * this file re-parses composer text: a second grammar is precisely the drift
 * `/pay` already suffered across four implementations.
 *
 * [mullu.comrade.voice.VoiceCommand] **still has its own grammar** and has not
 * been folded into core. That is a known follow-up, recorded in
 * `docs/CHAT_ACTIONS.md` §7 — not a claim that it already happened. It is a
 * refactor of a parser with its own passing tests, and was left alone rather
 * than done blind.
 */

/** What the composer should do once a command has been parsed. */
sealed interface ComposerPlan {
    /** Not a command — send the text as typed. */
    data object Send : ComposerPlan

    /** Route to [ComradeCore.taraAsideTyped]; never to a DM. `/tara`. */
    data class Aside(val text: String) : ComposerPlan

    /**
     * Route to [ComradeCore.taraInChatTyped] — `@tara`, which the peer sees.
     *
     * Separate from [Aside] because the audience is different, and a composer
     * that treated them the same would either publish a private thought or hide
     * an answer the other person was waiting for.
     */
    data class TaraHere(val text: String) : ComposerPlan

    /**
     * Two contacts answer to one `@handle`, so the command cannot run until the
     * user says which person they meant.
     *
     * Not an [Explain]: the old behaviour said "pick which one" and offered
     * nothing to pick, which is a dead end dressed as an instruction. The
     * composer shows [candidates] and re-runs the same draft once one is chosen
     * (see [withChoices]).
     */
    data class Choose(
        val handle: String,
        val candidates: List<ComradeCore.ContactInfo>,
    ) : ComposerPlan

    /** Route to [ComradeCore.assignTaskTyped]. [peer] null is a note to self. */
    data class Task(val text: String, val peer: String?) : ComposerPlan

    /**
     * File the thread containing [messageId] under [slug] —
     * [ComradeCore.assignThreadTyped].
     *
     * [messageId] is whatever the composer had selected, not the thread root:
     * core walks up the reply chain, so a frontend cannot file the wrong thread
     * by picking the wrong end of one.
     */
    data class AssignTopic(val slug: String, val messageId: String) : ComposerPlan

    /**
     * Open the topic sheet — a bare `/assign`, which is the discoverable way in
     * rather than a refusal.
     *
     * Not an [Explain]: "name a topic" with no way to see which topics exist is
     * the same dead end [Choose] was created to fix.
     */
    data object OpenTopics : ComposerPlan

    /** Route to [ComradeCore.offerActionTyped]. */
    data class Offer(val action: AppAction, val peers: List<String>) : ComposerPlan

    /** Open a screen on this device. */
    data class Open(val action: AppAction) : ComposerPlan

    /**
     * Start a `/play` flow with this query.
     *
     * [service] is what the *command* named (`/spotify …`), not what the query
     * turns out to be — a Spotify URL pasted after `/youtube` is still Spotify,
     * and `comrade_ui::play_query` is where that is settled.
     */
    data class Play(val query: String, val service: MusicService?) : ComposerPlan

    /** Show the command list. */
    data object Help : ComposerPlan

    /**
     * Cannot proceed, and here is the sentence to say. Covers both "this window
     * cannot do that" and "you have not said enough yet" — the composer treats
     * them the same way (say it, keep the text), so they are one case.
     */
    data class Explain(val message: String) : ComposerPlan
}

object ChatCommands {

    /**
     * Screens this Android build has.
     *
     * All five, unlike desktop — which is why the divergence is a named set on
     * both sides rather than an assumption of parity on either.
     */
    val AVAILABLE_SCREENS: Set<AppAction> = setOf(
        AppAction.BREATHE,
        AppAction.FOCUS,
        AppAction.JOURNAL,
        AppAction.TARA,
        AppAction.READ,
    )

    /**
     * What the composer should do with [command].
     *
     * [mentions] is the resolved form from [ComradeCore.resolveMentions], keyed
     * by handle. The one rule worth stating: **an unrecognised command never
     * sends.** A message silently swallowed is worse than one that arrives
     * confusing, because the sender believes they sent it.
     */
    fun planFor(
        command: ChatCommand,
        mentions: List<ComradeCore.MentionMatchInfo> = emptyList(),
        replyTarget: String? = null,
    ): ComposerPlan = when (command) {
        is ChatCommand.Plain -> ComposerPlan.Send

        // Recognised so it is never called a typo, then sent as text exactly as
        // it always has been — the chip and the peer's own client do the rest.
        is ChatCommand.Pay -> ComposerPlan.Send

        is ChatCommand.Help -> ComposerPlan.Help

        is ChatCommand.AskTara ->
            if (command.text.isBlank()) {
                ComposerPlan.Explain("Say what you want to think through — only you will see it.")
            } else {
                ComposerPlan.Aside(command.text.trim())
            }

        is ChatCommand.TaraHere ->
            if (command.text.isBlank()) {
                ComposerPlan.Explain("Ask her something — you'll both see her answer.")
            } else {
                ComposerPlan.TaraHere(command.text.trim())
            }

        is ChatCommand.Task -> {
            if (command.text.isBlank()) {
                ComposerPlan.Explain("What needs doing?")
            } else {
                when (val targets = resolve(command.assignees.map { it.handle }, mentions)) {
                    is Resolution.Problem -> ComposerPlan.Explain(targets.message)
                    is Resolution.Ambiguous ->
                        ComposerPlan.Choose(targets.handle, targets.candidates)
                    is Resolution.Ok ->
                        ComposerPlan.Task(command.text.trim(), targets.npubs.firstOrNull())
                }
            }
        }

        is ChatCommand.OfferTo -> {
            if (command.targets.isEmpty()) {
                ComposerPlan.Explain("Name the comrade you want to offer this to, like @ana.")
            } else {
                when (val targets = resolve(command.targets.map { it.handle }, mentions)) {
                    is Resolution.Problem -> ComposerPlan.Explain(targets.message)
                    is Resolution.Ambiguous ->
                        ComposerPlan.Choose(targets.handle, targets.candidates)
                    is Resolution.Ok -> ComposerPlan.Offer(command.action, targets.npubs)
                }
            }
        }

        is ChatCommand.Open ->
            if (command.action in AVAILABLE_SCREENS) {
                ComposerPlan.Open(command.action)
            } else {
                ComposerPlan.Explain("${labelFor(command.action)} isn't in this app yet.")
            }

        is ChatCommand.Play ->
            if (command.query.isBlank()) {
                ComposerPlan.Explain("Name a song, or paste a link.")
            } else {
                ComposerPlan.Play(command.query.trim(), command.service)
            }

        // The three-way answer is `TopicDecisions.planAssign`'s, shared with the
        // desktop and pinned by mirrored vectors — this only turns it into the
        // sentence Android says.
        is ChatCommand.AssignTopic ->
            when (
                val plan = mullu.comrade.topic.TopicDecisions.planAssign(
                    command.topics.map { it.slug },
                    replyTarget,
                )
            ) {
                is mullu.comrade.topic.TopicDecisions.AssignPlan.OpenPicker ->
                    ComposerPlan.OpenTopics
                is mullu.comrade.topic.TopicDecisions.AssignPlan.NeedsTarget ->
                    ComposerPlan.Explain(
                        "Reply to a message first — then /assign #${plan.slug} files that thread.",
                    )
                is mullu.comrade.topic.TopicDecisions.AssignPlan.File ->
                    ComposerPlan.AssignTopic(plan.slug, plan.messageId)
            }

        is ChatCommand.Unknown ->
            ComposerPlan.Explain("There is no /${command.name} — type / to see what there is.")
    }

    private sealed interface Resolution {
        data class Ok(val npubs: List<String>) : Resolution
        data class Ambiguous(
            val handle: String,
            val candidates: List<ComradeCore.ContactInfo>,
        ) : Resolution

        data class Problem(val message: String) : Resolution
    }

    /**
     * Turn handles into npubs, or say what stopped it.
     *
     * The ambiguous case is the one that matters: a handle is a self-declared
     * alias and two contacts can answer to one, so picking the first is how a
     * private message reaches the wrong person. It comes back as
     * [Resolution.Ambiguous] rather than an error, because the answer exists —
     * only the user has it.
     *
     * One handle at a time, deliberately: `/comrade-breathe @ana @ana` would have
     * two questions to ask, and asking them at once means a chooser that can be
     * half-answered. The first unresolved handle is returned and the next is
     * asked about on the re-run.
     */
    private fun resolve(
        handles: List<String>,
        mentions: List<ComradeCore.MentionMatchInfo>,
    ): Resolution {
        val npubs = mutableListOf<String>()
        for (handle in handles) {
            val match = mentions.firstOrNull { it.handle == handle }
            when {
                match == null || (match.npub == null && match.candidates.isEmpty()) ->
                    return Resolution.Problem("@$handle isn't in your contacts.")
                match.npub == null -> return Resolution.Ambiguous(handle, match.candidates)
                else -> npubs.add(match.npub)
            }
        }
        return Resolution.Ok(npubs)
    }

    /**
     * Apply the choices a user has made for ambiguous handles, so re-running the
     * same draft reaches the person they picked.
     *
     * A choice is honoured **only** when it names one of that handle's own
     * candidates. Without that check, a pin left over from an earlier draft — or
     * from before a contact was removed — would silently retarget a message,
     * which is the exact failure the ambiguity exists to prevent.
     */
    fun withChoices(
        mentions: List<ComradeCore.MentionMatchInfo>,
        chosen: Map<String, String>,
    ): List<ComradeCore.MentionMatchInfo> = mentions.map { match ->
        val pick = chosen[match.handle]
        if (match.npub == null && pick != null && match.candidates.any { it.npub == pick }) {
            match.copy(npub = pick)
        } else {
            match
        }
    }

    /**
     * How one candidate reads in the chooser.
     *
     * The key is not decoration here: two contacts answering to one handle very
     * often have the *same alias too* — that is why the handle is ambiguous — so
     * a row showing only the name would offer two identical choices. [title] is
     * the caller's resolved display name; the short key is what tells them apart.
     */
    fun candidateLabel(title: String, npub: String): String = "$title · ${shortNpub(npub)}"

    /**
     * The helplines as the composer can show them.
     *
     * `AUDIT.md` §8's honesty gate says the resources appear beside *any* reply
     * that tripped the distress detector — and Tara can now be asked from a chat,
     * where the only affordance is a single `Text`. So `TaraScreen.kt`'s
     * `CrisisCard` gets a plain-text twin rather than the gate quietly applying to
     * one screen. Empty in, empty out: a note must not end with a dangling
     * heading when core returned nothing.
     */
    fun crisisLines(resources: List<ComradeCore.CrisisResourceInfo>): String =
        if (resources.isEmpty()) {
            ""
        } else {
            "You don't have to carry this alone:\n" +
                resources.joinToString("\n") { "· ${it.name} — ${it.contact}" }
        }

    /**
     * What to say after a `/play`, given the route core decided.
     *
     * Pure and tested, because every one of these sentences is a claim about
     * what the app just did or can do, and three of the five are refusals. A
     * refusal that names the wrong reason is worse than none: it sends someone
     * looking for a file picker when the real problem is that Spotify will not
     * let another app decode its audio.
     *
     * Returns `null` for [PlayRoute.START_TOGETHER] handled with a title — the
     * caller says which recording opened, which it knows and this does not.
     */
    fun playNote(route: PlayRoute, service: MusicService?, query: String): String = when (route) {
        // No longer "it's under Together": the session overlay covers the app
        // the moment one exists, so it is already on screen by the time this is
        // read, and sending someone to look for it was sending them nowhere.
        PlayRoute.START_TOGETHER -> "Playing \"$query\" together."
        // The confidence bar did its job — opening the wrong track on someone's
        // behalf is worse than asking. The picker is already open by the time
        // this is read, so this says *why* it opened rather than what to do.
        PlayRoute.ASK_FOR_FILE ->
            "No copy of \"$query\" on this phone — pick the file."
        // Reworded 2026-08-08. The old sentence said the service "won't let
        // another app play its audio", and that was never quite the reason —
        // decoding their bytes is out, but driving playback inside their own
        // client on this listener's own subscription is exactly what their SDKs
        // are for. What is actually missing on this device is the connected
        // account, so that is what it now says. See docs/TOGETHER.md §11.
        PlayRoute.OPEN_ELSEWHERE ->
            "No ${serviceLabel(service)} account connected here — open it there."
        // The account is connected and the track can be driven: a real session,
        // playing from this listener's own subscription. Nothing is shared but
        // the playhead.
        PlayRoute.PLAY_ON_SERVICE ->
            "Playing \"$query\" together on ${serviceLabel(service)}."
        // Reworded 2026-08-08, when the embed landed. This used to refuse —
        // "Comrade can't play YouTube here" — because core could carry the
        // invitation and nothing on this side could play it. It can now, and the
        // sentence has to change with it or the app tells people it cannot do
        // the thing it just did (docs/TOGETHER.md §11b).
        //
        // The one route to "play something neither of us has" that needs no
        // account on either side, which is worth saying in the note: the query
        // is a link, so there is nothing about a local copy to explain.
        PlayRoute.PLAY_EMBED -> "Watching that together — neither of you needs a copy."
        PlayRoute.NOTHING -> "Name a song, or paste a link."
    }

    /**
     * What to say when the route is "no copy here" but this device was never
     * allowed to look.
     *
     * Separate from [playNote]'s `ASK_FOR_FILE` on purpose: that sentence says a
     * copy is not on the phone, and saying it when Comrade simply cannot read
     * the music library would send someone hunting for a file they already have.
     * `LibraryResolver.mayRead` is what distinguishes them.
     */
    const val LIBRARY_UNSEEN: String =
        "Comrade can't read your music library, so it can't look — pick the file."

    /** Human name for a music service, for [playNote]. */
    fun serviceLabel(service: MusicService?): String = when (service) {
        MusicService.SPOTIFY -> "Spotify"
        MusicService.APPLE_MUSIC -> "Apple Music"
        MusicService.YOUTUBE -> "YouTube"
        // A link we recognised as DRM-bearing but whose service did not survive
        // — better a true vague sentence than a confidently wrong brand.
        null -> "That service"
    }

    /** Human name for a destination, for the sentences above. */
    fun labelFor(action: AppAction): String = when (action) {
        AppAction.BREATHE -> "Taking a deep breath"
        AppAction.FOCUS -> "Focus sessions"
        AppAction.JOURNAL -> "The journal"
        AppAction.TARA -> "Tara"
        AppAction.READ -> "The reader"
    }

    /**
     * The `/` picker's rows for what has been typed so far, or an empty list when
     * the picker should be closed.
     *
     * Matches aliases as well as names, so typing `/listen` finds `play` — an
     * alias that works on submit but is invisible while typing reads as not
     * existing. Closes once there is a space, because what follows is an
     * argument rather than a name.
     */
    fun pickerRows(text: String, catalog: List<CommandSpec>): List<CommandSpec> {
        if (!text.startsWith("/")) return emptyList()
        if (text.any { it.isWhitespace() }) return emptyList()
        val typed = text.removePrefix("/").lowercase()
        return catalog.filter { spec ->
            spec.name.startsWith(typed) || spec.aliases.any { it.startsWith(typed) }
        }
    }

    /**
     * The text to put in the composer when a picker row is chosen.
     *
     * A trailing space when something follows, so the caret is already where the
     * next word goes; none when nothing does, so the send button submits instead
     * of sending `"/breathe "`.
     */
    fun completionFor(spec: CommandSpec): String =
        if (spec.argument.isNotEmpty() || spec.takesMention) "/${spec.name} " else "/${spec.name}"

    /**
     * Who will read what is in the composer, when Tara is being addressed.
     *
     * The sigil is the whole difference, so the composer has to show it *before*
     * the send button is pressed — a private thought that looks like a message,
     * or a question the other person turns out never to have seen, are both
     * failures of this one label.
     */
    enum class TaraAudience {
        /** `/tara` — never leaves the device. */
        Private,

        /** `@tara` — the question and her answer both go to the peer. */
        Shared,
    }

    /**
     * Which Tara audience the raw draft implies, or null if she is not addressed.
     *
     * Decided from the raw text rather than a parse, so it is right from the
     * moment `@tara ` is typed — before there is anything to parse. Deliberately
     * eager for the same reason it always was: mislabelling this is how somebody
     * sends a private thought by accident.
     */
    fun taraDraft(text: String): TaraAudience? {
        val t = text.trimStart()
        val lower = t.lowercase()
        for ((prefix, audience) in
            listOf("@tara" to TaraAudience.Shared, "/tara" to TaraAudience.Private)
        ) {
            if (lower == prefix) return audience
            if (lower.startsWith(prefix) && t.length > prefix.length &&
                t[prefix.length].isWhitespace()
            ) {
                return audience
            }
        }
        return null
    }
}
